examples/image_classifier/image_classifier.py

```diff
@@ -0,0 +1,263 @@
+            def __init__(self):
+                super().__init__()
+                self.conv1 = nn.Conv2d(3, 6, 5)
+                self.pool = nn.MaxPool2d(2, 2)
+                self.conv2 = nn.Conv2d(6, 16, 5)
+                self.fc1 = nn.Linear(16 * 5 * 5, 120)
+                self.fc2 = nn.Linear(120, 84)
+                self.fc3 = nn.Linear(84, 10)
+
+            def forward(self, x):
+                x = self.pool(F.relu(self.conv1(x)))
+                x = self.pool(F.relu(self.conv2(x)))
+                x = torch.flatten(x, 1)  # flatten all dimensions except batch
+                x = F.relu(self.fc1(x))
+                x = F.relu(self.fc2(x))
+                x = self.fc3(x)
+                return x
+
+        net = Net().to(device)
+
+        # Define a Loss function and optimizer
+        criterion = nn.CrossEntropyLoss()
+        optimizer = optim.SGD(net.parameters(), lr=0.001, momentum=0.9)
+
+        # Load train data
+        batch_size = 4
+        trainloader = torch.utils.data.DataLoader(
+            self.trainset, batch_size=batch_size, shuffle=True, num_workers=2
+        )
+
+        # Start training
+        num_epochs = 2
+        for epoch in range(num_epochs):  # loop over the dataset multiple times
+            running_loss = 0.0
+            for i, data in enumerate(trainloader, 0):
+                # get the inputs; data is a list of [inputs, labels]
+                inputs, labels = data[0].to(device), data[1].to(device)
+
+                # zero the parameter gradients
+                optimizer.zero_grad()
+
+                # forward + backward + optimize
+                outputs = net(inputs)
+                loss = criterion(outputs, labels)
+                loss.backward()
+                optimizer.step()
+
+                # print statistics
+                running_loss += loss.item()
+                if i % 2000 == 1999:  # print every 2000 mini-batches
+                    print(f"[{epoch + 1}, {i + 1:5d}] loss: {running_loss / 2000:.3f}")
+                    # log metrics to wandb
+                    wandb.log({"mini-batches": {i + 1}, "loss": {running_loss / 2000}})
+                    running_loss = 0.0
+
+        print("Finished Training")
+        buffer = BytesIO()
+        torch.save(net.state_dict(), buffer)
+        self.model_state_dict_bytes = buffer.getvalue()
+        self.next(self.evaluate)
+
+    # Test the model on the test data
+    @pypi(
+        python="3.11.9",
+        packages={
+            "torch": "2.4.1",
+            "torchvision": "0.19.1",
+        },
+    )
+    @kubernetes
+    @step
+    def evaluate(self):
+        import torch
+        import torch.nn as nn
+        import torch.nn.functional as F
+        from io import BytesIO
+
+        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+        print(f"Evaluating on: {device}")
+
+        # Define a Convolutional Neural Network
+        class Net(nn.Module):
```

Comment:
So, this part of the code does copy exactly the implementation of `Net` from the train step. I understand why this is the case (when you're hacking this is easier than figuring out imports while you're changing things and it's not too bad when the model is this simple), but were our clients to use this as the example they paste into, we could be looking at a lot more lines of code duped between steps, and that raises the risk of the implementations getting out of step with each other.

Conveniently with Metaflow flows, you can make a file that is a sibling to the flow and import stuff from that file like you would a normal python class. So I'd name this class `ImageConvNN` or sometihng and then put it in a file called `image_conv_nn.py`, and import it in each step that needs it with `from image_conv_nn import ImageConvNN`.

Reply:
This is exactly what I was looking for. Thanks :) I will make this change 👍🏾